src/sharing/SharingDialog.component.js

```diff
@@ -1,44 +1,251 @@
+              if (httpStatus === 'OK') {
+                  this.setState({
+                      objectToShare,
+                      apiObject,
+                  }, () => {
+                      onSuccess && onSuccess();
+                  });
+              } else {
+                  console.warn('Failed to post changes.');
+                  console.warn('SERVER SAID:', message);
+              }
 
-    mixins: [Translate],
+              return message;
+          })
+          .catch(({ message }) => {
+              console.warn('Failed to post changes.');
+              console.warn('SERVER SAID:', message);
+          });
+    }
+
+    loadObjectFromApi() {
+        getInstance().then((d2) => {
+            const apiInstance = d2.Api.getApi();
+            apiInstance.get('sharing', { type: this.props.type, id: this.props.id })
+                .then((apiObject) => {
+                    this.setState({
+                        api: apiInstance,
+                        apiObject,
+                        objectToShare: transformObjectStructure(apiObject.meta, apiObject.object),
+                    });
+                });
+        });
+    }
+
+    restoreObjectStructure(transformedObject) {
+        const userAccesses = [];
+        const userGroupAccesses = [];
+
+        // eslint-disable-next-line array-callback-return
+        transformedObject.accesses.map((access) => {
+            const apiAccess = {
+                id: access.id,
+                name: access.name,
+                displayName: access.name,
+                access: cachedAccessTypeToString(access.canView, access.canEdit),
+            };
+
+            access.type === 'user'
+                ? userAccesses.push(apiAccess)
+                : userGroupAccesses.push(apiAccess);
+        });
+
+        return {
+            meta: this.state.apiObject.meta,
+            object: {
+                ...this.state.apiObject.object,
+                ...userAccesses && { userAccesses },
+                ...userGroupAccesses && { userGroupAccesses },
+
+                publicAccess: cachedAccessTypeToString(
+                    transformedObject.publicCanView,
+                    transformedObject.publicCanEdit
+                ),
+                externalAccess: transformedObject.isSharedExternally,
+            },
+        };
+    }
+
+    closeSharingDialog() {
+        this.props.onRequestClose(this.state.apiObject.object);
+    }
 
     render() {
         const sharingDialogActions = [
             <FlatButton
-                label={this.getTranslation('close')}
-                onClick={this.closeSharingDialog} />,
+                label={this.context.d2.i18n.getTranslation('close')}
+                onClick={this.closeSharingDialog.bind(this)}
+            />,
         ];
 
-        return (
+        const loadingMaskStyle = {
+            position: 'relative',
+        };
+
+        return this.state.objectToShare ? (
```

Comment:
I know this is a remnant of my old stuff and that we have this in multiple places but perhaps we should rewrite it to be like the following to make it somewhat cleaner?

```js
if (this.state && !this.state.objectToShare) {
    return (<LoadingMask style={loadingMaskStyle} size={1} />); 
}

return <Dialog ... />
```